File: WHO all groups/wHV/goveqs_basis2.m
function [out, lam] = goveqs_basis(t, in, M, i, prm, sel, agg)             % <--- Arguments changed to accommodate prm

invec = in(1:i.nstates);

if t<1980                                                                  
    rHIV = 0;
else
    rHIV = interp1(0:length(prm.rHIV)-1, prm.rHIV, (t-1980)); 
end
    
try
% Normalise by populations
lam = M.lambda*invec/sum(invec);
allmat = M.lin + M.Dxlin + rHIV*M.linHIV + lam*M.nlin;                     
out = allmat*invec;
catch
   keyboard; 
end

% Implement deaths
morts = sum(M.mortvec,2).*invec;
out = out - morts;

% Implement births
births = sum(morts);
out(i.U.v0.h0) = out(i.U.v0.h0)+births;

% Get the auxiliaries
out(i.aux.inc)  = agg.inc*(sel.inc.*allmat)*invec;
out(i.aux.noti) = agg.noti*(sel.noti.*allmat)*invec;
out(i.aux.mort(1)) = sum(M.mortvec(:,2).*invec);
out(i.aux.mort(2)) = sum(M.mortvec(:,3).*invec);
out(i.aux.Dx)      = agg.Dx*(sel.Dx.*allmat)*invec;

File: Python/goveqs_basis2.py
import numpy as np
from scipy.sparse import csr_matrix
import pdb
from setup_model import *

def goveqs_basis2(t, in_vec, i, s, M, agg, sel, r, p, prm):


    out = np.zeros(i['nx'])
    invec = in_vec[: i['nstates']]
    invec = np.array(invec, dtype=float)

    # Prepare population denominators
    tmp = M['denvec'] @ invec
    den = np.sum(M['denvec'].toarray() * tmp[:, None], axis=0)
    den[den == 0] = np.inf

    if t < 1980:
        rHIV_val = 0
    else:
        rHIV_val = np.interp(t - 1980, np.arange(len(prm['rHIV'])), prm['rHIV'])

    try:
        lam_val = (M['lam'] @ (invec / den)) * ((1 - p['betadec']) ** max(t - 2010, 0))
        allmat = (M['lin'] +
                  rHIV_val * M['linHIV'] +
                  lam_val[0] * M['nlin']['ch'] +
                  lam_val[1] * M['nlin']['ad'])
        allmat = np.array(allmat, dtype=float)
        out = allmat @ invec
    except Exception as e:
        pdb.set_trace()
    
    
    
    # Mortality and births
    morts = M['mort'].toarray() * invec.reshape(-1, 1)
    out[: i['nstates']] = out[: i['nstates']] - np.sum(morts, axis=1)
    
    dom_morts = np.sum(morts[np.array(s['dom']) - 1, :])
    out[i['U']['ch']['dom']['neg'] - 1] = out[i['U']['ch']['dom']['neg'] - 1] + dom_morts
    
    # Auxiliaries
    out[np.array(i['aux']['inc']) - 1] = agg['inc'] @ (np.multiply(sel['inc'], allmat) @ invec)
    out[np.array(i['aux']['mort']) - 1] = np.sum(morts[:, 1])
    out[np.array(i['aux']['nTPT']) - 1] = np.sum((np.multiply(sel['nTPT'], allmat) @ invec))
    out[np.array(i['aux']['ch_notifs']) - 1] = np.sum((np.multiply(sel['ch_notifs'], allmat) @ invec))

    return out


